Adding of missing wavelets from the matlab list

holgern opened the pull request:
* Missing wavelets are added #165
* 1d cwt added (not complete)  #48

holgern commented:
This is not finished by now.
What do you think? Should i proceed or did i make some design mistakes?

grlee77 commented on pywt/_extensions/_cwt.pyx:
```diff
@@ -0,0 +1,108 @@
+
+
+cpdef cwt_conv(data_t[::1] data, data_t[::1] in_filter, size_t output_len):
+    cdef size_t N, F, O
```
When building I got:
`warning: _cwt.pyx:71:18: Index should be typed for more efficient access`
It seems the index, `i`, used in `cwt_conv` needs to be included in the list of `size_t` cdefs

grlee77 commented on pywt/_extensions/wavelets_list.pxi:
```diff
@@ -5,128 +5,156 @@
-    "coif7": (c"c", 7),
-    "coif8": (c"c", 8),
-    "coif9": (c"c", 9),
+    "coif1": (4, 1),
```
For clarity, it might be worth also including the WAVELET_NAME ctypedef in this file as was done in wavelet.pxd, so that these would be:
```"coif2": (COIF, 2), ```
etc.
That would also keep this integers from getting out of sync if the enum in c/wavelets.h changes for some reason in the future.

grlee77 commented:
@kwohlfahrt has the most familiarity with the current c extensions, but this looks pretty reasonable to me.  I have not checked the implementation of the algorithm at all at this point, but the organization seems reasonable.

Are there particular design choices you would like feedback on?  From what I can tell, it seems the following properties were added to the Wavelet object:

`dwt_possible`, `cwt_possible`, `complex_cwt`, `lower_bound`, `upper_bound`.  Of these, most are specific to continuous wavelets.  It looks like `lower_bound` and `upper_bound` will return `None` for discrete wavelets which seems reasonable.  Maybe someone with more object-oriented programming experience can comment on whether from an API standpoint, it is preferable to having a separate ContinuousWavelet class that adds those properties?

grlee77 commented:
Also, how does this compare to the CWT functionality in scipy?
http://docs.scipy.org/doc/scipy-0.17.0/reference/generated/scipy.signal.cwt.html

or to that in @aaren 's python-based package?:
https://github.com/aaren/wavelets

holgern commented:
Thank you @grlee77 for the review.
What are new and not implemented are the wavelet funktions. I plan also to be able to use all DWT-Wavelets for cwt (e.g. db4). 
I added now all remaining wavelets except the Meyer wavelet
We can exchange the CWT function and use either scipy or the one from  @aaren 
At the moment, the cwt function i implemented is slow

kwohlfahrt commented on pywt/_extensions/_pywt.pyx:
```diff
@@ -397,6 +399,67 @@ cdef public class Wavelet [type WaveletType, object WaveletObject]:
+        def __set__(self, float value):
+            self.w.upper_bound = value
+
+    property fc:
```
Spell out properties. `center_frequency` is much clearer than `fc`.

kwohlfahrt commented on pywt/_extensions/c/cwt.template.c:
```diff
@@ -0,0 +1,343 @@
+int CAT(TYPE, _gaus)(const TYPE * const restrict input,
+                              TYPE * const restrict output, const size_t N,
+                              const size_t number){
+    //if (strcmp(wavelet->short_name, "gauss")){
```
No commented out code please.

kwohlfahrt commented on pywt/_extensions/c/cwt.template.c:
```diff
@@ -0,0 +1,343 @@
+    //if (strcmp(wavelet->short_name, "gauss")){
+           switch (number) {
+               case 1:
+                    CAT(TYPE, _gaus1)(input, output, N);
```
I would inline the calculation, and let the compiler unswitch it. So like:

```
for (i = 0; i < N; i++) {
  switch(number) {
    ...
  }
}
```

This should make things a lot more compact and readable, and any sensible compiler should optimize this correctly.

kwohlfahrt commented on pywt/_extensions/_cwt.pyx:
```diff
@@ -0,0 +1,156 @@
+
+    #if data_t is np.float64_t:
+        # TODO: Don't think these have to be 0-initialized
+        # TODO: Check other methods of allocating (e.g. Cython/CPython arrays)
```
Yes, we should do this for other functions as well.

kwohlfahrt commented on pywt/_extensions/c/cwt.template.c:
```diff
@@ -0,0 +1,343 @@
+
+
+
+int CAT(TYPE, _gaus)(const TYPE * const restrict input,
```
None of these functions return anything but `0`, so they can be `void`.

kwohlfahrt commented on pywt/_extensions/c/wavelets.c:
```diff
@@ -328,36 +519,35 @@ Wavelet* blank_wavelet(size_t filters_length)
-       w->rec_lo_double == NULL || w->rec_hi_double == NULL){
-        free_wavelet(w);
-        return NULL;
+    if (filters_length > 0)
```
This does not set the filter pointers to `NULL` if `filters_length == 0`. That should be done, or the allocation can be removed from the `if` statement (`malloc(0)` is allowed, but may return `NULL` so we should not check it afterwards).

kwohlfahrt commented:
Well, We've got enough requests for it that I think we should add the CWT. So I think this is a good plan in general. I might not have much time to review this until next Monday, but I've had a quick look. If I've missed something let me know please!

I have made a few in-line comments, my main criticism from a design point of view is what @grlee77 said. The new continuous wavelets and the discrete wavelets don't really have much overlap in their fields. Could you see what is possible to separate out? Both in terms of the C struct and the Python/Cython object.

holgern commented:
For the meyer wavelet, we neeed ifft. This meens more dependencies. Should i add it?

holgern commented:
Why is assert_allclose(np.imag(PSI), np.imag(psi)) failing for linux? It works for me (Windows x64)

grlee77 commented:
In regards to the test failures, all of the tests in your PR are passing on my linux machine.  Looking at the Travis output for the failures, it seems like the results are very close, it is probably just a numerical precision issue causing it to not quite meet the default tolerances in `assert_allclose`.  Maybe try bumping up rtol and/or atol slightly?

As far as adding FFT code, I assume you want this for use in the C or Cython code?  If so, I think that would require more discussion of the appropriate approach and would probably best fit in a separate issue/PR.  Maybe just omit the Meyer wavelet in the present PR for simplicity.  Alternatively, if you can do the FFTs from python you can just use numpy.fft without needing to add any new dependencies.

kwohlfahrt commented on pywt/_extensions/c/cwt.template.h:
```diff
@@ -0,0 +1,66 @@
+                              TYPE * const restrict output, const size_t N,
+                              const size_t number);
+
+void CAT(TYPE, _gaus1)(const TYPE  input, TYPE * const restrict output);
```
If not inlined, I think it's better to keep these as private functions (i.e. don't exposre them in the header), unless you can think of a reason they would be used elsewhere?

kwohlfahrt commented on pywt/_extensions/c/cwt.template.c:
```diff
@@ -0,0 +1,267 @@
+    }
+}
+
+void CAT(TYPE, _gaus1)(const TYPE  input, TYPE * const restrict output)
```
These aren't used anywhere else, so I would inline them into `TYPE_gaus` (unless I'm missing something?) It'd make things a bit more compact.

kwohlfahrt commented on pywt/_extensions/c/wavelets.h:
```diff
@@ -13,6 +13,24 @@ typedef enum {
 } SYMMETRY;
 
+/* Wavelet name */
+typedef enum {
```
I like this idea. I don't think there's a need to assign numbers, `enum {HAAR, REVERSE_BIORTHOGONAL, ...}` should be fine.

kwohlfahrt commented on pywt/_extensions/_pywt.pyx:
```diff
@@ -397,6 +399,67 @@ cdef public class Wavelet [type WaveletType, object WaveletObject]:
+        def __get__(self):
+            return bool(self.w.cwt_possible)
+        def __set__(self, int value):
+            self.w.cwt_possible = (value != 0)
```
I think this class should be refactored to separate out CWT and DWT wavelets. Something like this:

```
class Wavelet:
  # common properties
  ...

class DiscreteWavelet(Wavelet):
  # DWT specific things
  ...

class ContinuousWavelet(Wavelet):
  # CWT specific things
  ...
```

If a method is available for both classes, but can't be defined without knowing whether the class is continuous or discrete, define it on `Wavelet`, but use `abc.abstractmethod` if it's available in Cython, or raise `NotImplementedError` otherwise.

Hopefully this will show what can be split over two separate C structs as well.

I realize this could be a bit complicated so I'm happy to answer any questions. I'm able to respond to things more quickly again now so it shouldn't take too long.

kwohlfahrt commented on pywt/_extensions/c/wavelets.c:
```diff
@@ -291,7 +314,14 @@ Wavelet* wavelet(char name, unsigned int order)
             w->short_name = "dmey";
-
+            w->dwt_possible = 1;
+            w->cwt_possible = 1;
```
Again, this would be nicer if it was split over two different structs. That way it'd be possible to catch some errors at compile time if the wrong type is used.

holgern commented:
It is not so easy to create to seperate structs for cwt and dwt.
The table Wavelet Families and Associated Properties  from http://de.mathworks.com/help/wavelet/ug/wavelet-families-additional-discussion.html?refresh=true 
shows this. 
Maybe we need three structs in wavelet.h?
/* Wavelet structure holding pointers to filter arrays and property attributes */
typedef struct {
    /* Wavelet properties */
    int vanishing_moments_psi;
    int vanishing_moments_phi;
    index_t support_width;

    SYMMETRY symmetry;

    unsigned int orthogonal:1;
    unsigned int biorthogonal:1;
    unsigned int compact_support:1;
    
    unsigned int dwt_possible:1;
    unsigned int cwt_possible:1;
    
    float lower_bound;
    float upper_bound;    

    char* family_name;
    char* short_name;
    DiscreteWavelet* discrete_wavelet;
    ContinuousWavelet* continous_wavelet;

} Wavelet;

typedef struct {
    double* dec_hi_double;  /* highpass decomposition */
    double* dec_lo_double;  /* lowpass decomposition */
    double* rec_hi_double;  /* highpass reconstruction */
    double* rec_lo_double;  /* lowpass reconstruction */
    float* dec_hi_float;
    float* dec_lo_float;
    float* rec_hi_float;
    float* rec_lo_float;

    size_t dec_len;   /* length of decomposition filter */
    size_t rec_len;   /* length of reconstruction filter */


} DiscreteWavelet;

typedef struct {

    
    /* Parameters for shan, fbsp, cmor*/
    float center_frequency;
    float bandwidth_frequency;
    unsigned int fbsp_order;

} ContinuousWavelet;

kwohlfahrt commented:
Perhaps, though with the example you showed I would do it the other way around. Instead of:

```
struct Wavelet {
  unsigned int cwt_possible:1;
  unsigned int dwt_possible:1;
  DiscreteWavelet* discrete_wavelet;
  ContinuousWavelet* continuous_wavelet;
}

struct DiscreteWavelet {...};
struct ContinuousWavelet {...}:
```

Perhaps the following:

```
struct BaseWavelet {...};
struct DiscreteWavelet {
  BaseWavelet base;
  ...
};
struct ContinuousWavelet {
  BaseWavelet base;
  ...
};
```

This makes sure that the type hierarchy is defined by which struct a function takes, and not a boolean flag (`xyz_possible`) that is checked at runtime. Hopefully that will let the compiler catch more errors, and it'll mirror the inheritance used in the Python classes better too.

For functions that can take either a continuous or discrete wavelet, I suggest the dispatching is done in the Cython layer. Shared data that can use the same code should operate on the base wavelet struct, and if the code is significantly different separate functions are nicer anyway. We could use C11 generic functions, except we need to support MSVC.

PS apologies for the long absence, I'm writing some papers a thesis and applications all at the same time and it's a bit hectic.

rgommers commented:
> PS apologies for the long absence, I'm writing some papers a thesis and applications all at the same time and it's a bit hectic.

No need to apologize (and I'm sure I've been missing in action for longer .....), and good luck with all of those!

holgern commented:
What do you think? It is not finished, but a first start.

grlee77 commented:
This looks like a good start.  I would remove `dwt_possible` and `cwt_possible` from the `BaseWavelet` struct in `wavelets.c`.  I don't think these properties are needed anymore.  It seems like `wavelet.is_discrete_wavelet` should be sufficient to determine if the wavelet is discrete or not.

Similarly, do you really need separate `.dw` and `.cw` properties in the Wavelet object of `_pywt.pyx`?  It seems much simpler to just keep a `.w` property and in the few places where the code path is actually different, use `is_discrete_wavelet` to choose the appropriate branch.

holgern commented:
For image support, we need 2D cwt. The cwt convention function has to be enhanced to 2D.

holgern commented:
If we want to be similar to matlab than 2d-cwt has to be done by a new function cwtft2
mathworks.com/help/wavelet/ref/cwtft2.html
There is also a new set of wavelet supported:

```
Wavelet name	Parameters
'morl'	{'Omega0',6;'Sigma',1;'Epsilon',1}
'mexh'	{'p',2;'sigmax',1;'sigmay',1}
'paul'	{'p',4}
'dog'	{'alpha',1.25}
'cauchy'	{'alpha','pi/6';'sigma',1;'L',4;'M',4}
'escauchy'	{'alpha','pi/6';'sigma',1;'L',4;'M',4}
'gaus'	{'p',1;'sigmax',1;'sigmay',1}
'wheel'	{'sigma',2}
'fan'	{'Omega0X',5.336;'Sigma',1;'Epsilon',1;'J',6.5}
'pethat'	None
'dogpow'	{'alpha',1.25;'p',2}
'esmorl'	{'Omega0',6;'Sigma',1;'Epsilon',1}
'esmexh'	{'Sigma',1;'Epsilon',0.5}
'gaus2'	{'p',1;'sigmax',1;'sigmay',1}
'gaus3'	{'A',1;'B',1;'p',1;'sigmax',1;'sigmay',1}
'isodog'	{'alpha',1.25,1.25}
'dog2'	{'alpha',1.25,1.25}
'isomorl'	{'Omega0',6;'Sigma',1}
'rmorl'	{'Omega0',6;'Sigma',1;'Epsilon',1}
'endstop1'	{'Omega0',6}
'endstop2'	{'Omega0',6;'Sigma',1}
'gabmexh'	{'Omega0',5.336;'Epsilon',1}
'sinc'	{'Ax',1;'Ay',1;'p',1;'Omega0X',0;'Omega0Y',0}
```

So, what is missing for this request? Should we add 2D? 
We need no think about the convolution. We can use scipy.signal.convolve for 2D-Convolution as it is used by scikit-image. So if we want 2d-cwt we should use scipy.signal. Is this ok?

grlee77 commented:
I think it will be simpler to make 2D a separate, future PR.

holgern commented:
I almost finished now.
`[coef, freq]=pywt.cwt(data),np.arange(1,11),'db3',0.001)`
as well as 
`[coef, freq]=pywt.cwt(data),np.arange(1,11),'gaus1',0.001)`
is now possible.
The output of cwt is exactly the same as the matlab output.

holgern commented:
Why is appveyor for python 3.4 failing? 
```
[00:02:57] alloc: invalid block: 03C1DE18: 40 3
[00:03:00] Command exited with code -2147483645
```
Anyone has a idea?

rgommers commented:
> Why is appveyor for python 3.4 failing? 

This is very likely related to the changes to the build system made in gh-161. Note that the tests don't seem to be run on Appveyor for Python 2.7 and 3.5, so it's not just 3.4 that fails.

rgommers commented:
This PR is massive. Nice going @holgern

rgommers commented:
Question about the API of the functions added. Do you like the Matlab style, and do we want/need to keep following it in all cases? 

Something like ``cwt(data, scales, wavelet, sampling_period=1.)`` looks good, but signatures like ``cmorwavf(lb,ub,n,fb,fc)`` are about as unreadable as they come ....

kwohlfahrt commented:
Would it be possible to again separate the (python) class into a continuous and discrete wavelet class (as a sub-class of a base wavelet class)? Then the various `*wavf` functions can live on the continuous wavelet class.

kwohlfahrt commented on pywt/_extensions/_pywt.pxd:
```diff
@@ -8,7 +8,8 @@ ctypedef fused data_t:
 cdef public class Wavelet [type WaveletType, object WaveletObject]:
-    cdef wavelet.Wavelet* w
+    cdef wavelet.DiscreteWavelet* dw
+    cdef wavelet.ContinuousWavelet* cw
```
If we split `Wavelet` into `ContinuousWavelet` and `DiscreteWavelet`, then this would be much tidier. The `if dw != NULL` checks below would also all fall away, since those properties would only make sense on the `DiscreteWavelet`.

kwohlfahrt commented on pywt/_extensions/_pywt.pyx:
```diff
@@ -317,137 +328,279 @@ cdef public class Wavelet [type WaveletType, object WaveletObject]:
         "Lowpass reconstruction filter"
         def __get__(self):
-            return float64_array_to_list(self.w.rec_lo_double, self.w.rec_len)
+            if self.dw is not NULL:
```
See note above.

kwohlfahrt commented on pywt/_extensions/c/cwt.template.c:
```diff
@@ -0,0 +1,227 @@
+                 break;
+             case 8:
+                 output_r[i] = (256*(input[i]*input[i]*input[i]*input[i]*input[i]*input[i]*input[i]*input[i])*CAT(TYPE, _cos)(input[i])+1024*(input[i]*input[i]*input[i]*input[i]*input[i]*input[i]*input[i])*CAT(TYPE, _sin)(input[i])-5376*(input[i]*input[i]*input[i]*input[i]*input[i]*input[i])*CAT(TYPE, _cos)(input[i])-12544*(input[i]*input[i]*input[i]*input[i]*input[i])*CAT(TYPE, _sin)(input[i])+2 ...
+                 output_i[i] = (-256*(input[i]*input[i]*input[i]*input[i]*input[i]*input[i]*input[i]*input[i])*CAT(TYPE, _sin)(input[i])+1024*(input[i]*input[i]*input[i]*input[i]*input[i]*input[i]*input[i])*CAT(TYPE, _cos)(input[i])+5376*(input[i]*input[i]*input[i]*input[i]*input[i]*input[i])*CAT(TYPE, _sin)(input[i])-12544*(input[i]*input[i]*input[i]*input[i]*input[i])*CAT(TYPE, _cos)(input[i])- ...
```
Is there any reason for not using `pow(input[i], ...)` here?

kwohlfahrt commented on pywt/_extensions/c/wavelets.c:
```diff
@@ -234,19 +268,18 @@ Wavelet* wavelet(char name, unsigned int order)
-            w->short_name = "bior";
-
+            w->base.vanishing_moments_psi = order/10;
+            w->base.vanishing_moments_phi = -1;
```
This is the only discrete wavelet with `-1` for a vanishing moment. Is there a reason it doesn't make sense to define vanishing moments for the scaling function on bior wavelets?

kwohlfahrt commented on pywt/_extensions/c/wavelets.c:
```diff
@@ -316,123 +349,342 @@ Wavelet* wavelet(char name, unsigned int order)
+            w = blank_continous_wavelet();
+            if(w == NULL) return NULL;
 
+            w->base.vanishing_moments_psi = -1;
```
These attributes should probably live on the discrete wavelet struct, since they are not define for any continuous wavelet. Thoughts?

kwohlfahrt commented:
I've left a couple inline comments, but I think the big thing to decide is whether to split the discrete and continuous wavelet python classes.

Apart from that, I second what Ralf said. Massive effort, good work :)

holgern commented:
OkI splitted the Wavelet into ContinuousWavelet and DiscreteWavelet. What do you think?

kwohlfahrt commented on pywt/_extensions/_pywt.pyx:
```diff
@@ -233,27 +241,41 @@ def families(int short=True):
 
 
-cdef public class Wavelet [type WaveletType, object WaveletObject]:
+def Wavelet(name=u"", object filter_bank=None):
```
Changing `pywt.Wavelet` to a function instead of a class is useful because you can create both discrete wavelets and continuous wavelets from the name.

On the other hand, it's also an API break because `isinstance(w, pywt.Wavelet)` will no longer work. It might be best to have `Wavelet` be simple an alias to `DiscreteWavelet` (at least for now, perhaps deprecate it?), and then require explicit use of `ContinuousWavelet(name)`? I'm not sure...

kwohlfahrt commented on pywt/_extensions/_pywt.pyx:
```diff
@@ -372,40 +394,74 @@ cdef public class Wavelet [type WaveletType, object WaveletObject]:
+    property cwt_possible:
+        "CWT possible"
+        def __get__(self):
+            return True
```
These should no longer be necessary, we can just check the class. I don't think they are used outside of tests now.

kwohlfahrt commented on pywt/_extensions/_pywt.pyx:
```diff
@@ -372,40 +394,74 @@ cdef public class Wavelet [type WaveletType, object WaveletObject]:
+            if self.w.base.lower_bound != self.w.base.upper_bound:
+                return self.w.base.lower_bound
+        def __set__(self, float value):
+            if self.w is not NULL:
```
I don't think `self.w` can ever be `NULL` - this is just from a quick skim though, correct me if I'm wrong here!

kwohlfahrt replied:
Second, are the `upper_bound` and `lower_bound` properties ever used on the discrete wavelets? As far as I can tell they're just set to `+/- 5`, is there a reason for this value?

holgern commented:
Ok, I renamed DiscreteWavelet to Wavelet, so that is isinstance(w, pywt.Wavelet) works again.
The function DiscreteContinuousWavelet gives now either a Wavelet class for discrete wavelet or a ContinuousWavelet class for Continuous wavelets.

How it is possible to mark isinstance(w, pywt.Wavelet) as obsolete? Such that isinstance(w, pywt.ContinuousWavelet) or isinstance(w, pywt.DiscreteWavelet) should be used in the future?

holgern commented:
I think I will remove the various "morlet", "gauswavf", "mexihat","cmorwavf", "shanwavf", "fbspwavf", "cgauwavf" functions.  
There are all using wavefun from ContinuousWavelet.
For example the gauswavf function just doing:
```
    wavelet.upper_bound = ub
    wavelet.lower_bound = lb
    psi, x = wavelet.wavefun(length=n)
```
Do you agree?

grlee77 commented on pywt/_extensions/_pywt.pyx:
```diff
@@ -794,28 +798,42 @@ cdef public class ContinuousWavelet [type ContinuousWaveletType, object Continuo
+                x64 = np.linspace(self.w.lower_bound, self.w.upper_bound, output_length)
+            else:
+                x = np.linspace(self.w.lower_bound, self.w.upper_bound, output_length)
+                x32 = np.asarray(x, dtype=np.float32)
```
I was going to suggest passing the dtype to `np.linspace`, but it looks like that was only added in `numpy` 1.9.  I have opened #189 to discuss potentially bumping up our minimum numpy version.

grlee77 commented on pywt/_extensions/_pywt.pyx:
```diff
@@ -734,22 +734,63 @@ cdef public class ContinuousWavelet [type ContinuousWaveletType, object Continuo
+        -----
+        The effective support are set with lower_bound and upper_bound
+        The wavelet function is complex for cmor, shan, fbsp and cgau.
+        The  complex frequency B-spline  wavelet (fbsp) has bandwidth_frequency, center_frequency and fbsp_order as addional parameter
```
Change double to single space at "The  complex"  and  "B-spline  wavelet"

grlee77 commented:
I took a look again at the several most recent commits.  I appreciate that a lot of work has gone into this PR and think it is a great asset to `PyWavelets`!  I think the bottleneck has just been the amount of time we have had available as reviewers.

From what I can see, the recent changes look good and are in line with the previous feedback.  I made only a couple of minor comments.

grlee77 commented:
Also, I think it should be fine to declare all of the new functions imported from `cwt.h` in `c_wt.pxd` as `nogil` to be in line with the changes proposed for the discrete transforms in #188.

grlee77 commented on pywt/_cwt.py:
```diff
@@ -0,0 +1,90 @@
+    coefs : array_like
+        Continous wavelet transform of the input signal for the given scales and wavelet
+    frequencies : array_like
+        if the unit of sampling period are seconds and given, than frequencies are in hertz. Otherwise Sampling period of 1 is assumed.
```
I just sent a PR to your master branch containing some PEP8 related style fixes I made while reviewing the code.  That was easier than line commenting individually.

grlee77 commented:
@holgern
This needs a rebase after #188 was merged.  Ideally, please also add `nogil` to the new functions added to c_wt.pxd in this PR.

@PyWavelets/core  Does anyone else have additional feedback for this PR?  If not I will plan to merge next week.

holgern commented:
I will commit the change regarding passing the dtype to np.linspace when numpy is bumped to 1.9

grlee77 commented:
sounds good.  If the numpy 1.9 PR is merged prior to this one i will ping you.  

The nogil changes here look good to me.

rgommers commented:
that 1.9 PR is merged now

rgommers commented on pywt/_extensions/_pywt.pyx:
```diff
@@ -233,16 +241,30 @@ def families(int short=True):
     return __wfamily_list_long[:]
 
 
+def DiscreteContinuousWavelet(name=u"", object filter_bank=None):
```
This is a simple convenience utility with no docstring, but it looks like you're making it public. That doesn't look quite right.

aaren replied:
I feel that if anything the convenience method should just be called `Wavelet`, with `DiscreteWavelet` and `ContinuousWavelet` if you know what you want.

grlee77 replied:
@rgommers, @aaren 
I think this is the only other outstanding issue prior to merging this PR. @holgern has added a docstring since the original comment.

I agree that the naming suggested by @aaren is preferable, but the difficulty is that both `pywt` and likely a substantial amount of 3rd party code have statements like 

```python
if isinstance(wave, Wavelet):
```
that would be broken by switching the `Wavelet` class name to `DiscreteWavelet`. 

Unless there are strong objections to leaving the naming as is, I am inclined to merge as is.  The renaming and any potential deprecations, etc, can always be addressed in a separate future PR.

rgommers replied:
Agree that we can't really rename ``Wavelet`` without affecting a lot of user code. So okay with me to leave as is.

rgommers commented on pywt/_extensions/_pywt.pyx:
```diff
@@ -574,6 +607,255 @@ cdef public class Wavelet [type WaveletType, object WaveletObject]:
+
+    In order to use a built-in wavelet the parameter name must be
+    a valid name from the wavelist() list.
+
```
this is the main docstring where users will start, so would be good to add ``Attributes`` and ``Methods`` sections (in format like [here](https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt)), and note that the ``wavefun`` method has usage info/examples.

rgommers commented on pywt/data/_readers.py:
```diff
@@ -137,3 +137,43 @@ def ecg():
+
+def nino():
+    """
+    NINO3 sea surface temperature 1871-1996 (in C; from the UKMO GISST2.3):
```
insert a blank line here. ``autodoc`` formatting makes tables and should only grab the summary line.

rgommers replied:
More importantly, looks like this dataset doesn't have an appropriate license: http://www.metoffice.gov.uk/hadobs/gisst/terms_and_conditions.html

Or is there a public domain or more liberally licensed version somewhere else? http://paos.colorado.edu/research/wavelets/wave_python/ doesn't really say anything about that.

grlee77 replied:
If we cannot resolve the license, we can always replace the data in the demo with a different, public-domain timeseries.
There are various physiological recordings available here:

aaren replied:
It would be nice to use the nino stuff if possible. Torrence and Compo is the canonical cwt paper. I don't know much about this, but I wonder if example usage is ok (does fair use apply to small amounts of data?).

@pelson: maybe you can help us here - can you comment on whether UKMO has precedent for licensing out small amounts of data for example use in open source? Or is it a bit tricky?

rgommers commented on pywt/data/_readers.py:
```diff
@@ -137,3 +137,43 @@ def ecg():
+    time : ndarray
+       convenient timeseries to use for testing and demonstration
+    sst : ndarray
+       convenient timeseries to use for testing and demonstration
```
Maybe start with ``Sea surface temperature, ``

rgommers commented:
Overall looks quite good. The licensing of the dataset and whether ``DiscreteContinuousWavelet`` should be public are my only critical comments.

aaren commented:
Wow this took off! Great work @holgern :smile:

holgern commented:
I renamed DiscreteWavelet to Wavelet, so that isinstance(w, pywt.Wavelet) is not broken.
I will add documentation for DiscreteContinuousWavelet.

For future releases, DiscreteContinuousWavelet should be renamed to Wavelet and Wavelet should be renamed to DiscreteWavelet.

I will investigate the licensing of the dataset.

grlee77 commented on pywt/_extensions/_pywt.pyx:
```diff
@@ -574,6 +607,255 @@ cdef public class Wavelet [type WaveletType, object WaveletObject]:
+
+        Calculates approximations of scaling function (`phi`) and wavelet
+        function (`psi`) on xgrid (`x`) at a given level of refinement.
+
```
remove reference to phi in this docstring  (looks like a copy/paste error from the Discrete case)

holgern commented:
The nino3 data set is also available under:
http://hhstokes.people.uic.edu/ftp/b34sbin/b34sfiles/wavedata.mac
https://github.com/aaren/wavelets/blob/master/tests/nino3data.asc

The dataset is also used in this paper:
Torrence, C., and G. P. Compo, 1998: A practical guide to wavelet analysis. Bull. Amer. Meteor. Soc., 79, 6178 :
http://journals.ametsoc.org/doi/pdf/10.1175/1520-0477%281998%29079%3C0061%3AAPGTWA%3E2.0.CO%3B2

grlee77 commented:
Hi @holgern,  I think aside from the nino3 license issues this is good to go.  The original sea surface temperature data are avaialble from http://catalogue.ceda.ac.uk/uuid/fda780a7697f384010674ebb1d5e4f23 with full license terms here (here) [http://licences.ceda.ac.uk/image/data_access_condition/ukmo_agreement.pdf].  They specifically state


>UKMO data / software so obtained may be used solely for the purpose for which they were supplied. They may not be used for any other projects unless specific prior permission has been obtained in writing from the UKMO by a NERC Data Centre. Note that this applies even for other bona fide academic work.
>
>Data sets must not be passed on to third parties under any circumstances.


Have you tried requesting this permission?  Or could perhaps verify with one of the other sources linked to that they have obtained this permission previously?  As this is a tiny fraction (derived from?) the full dataset, I think it is fairly likely you could get permission, but I am not sure how long it might take.  I am also willing to write them to request this if you would like.

grlee77 commented:
Another possibility I would be okay with is to merge this PR without the example for now and add the example through a separate PR once the licensing issues are resolved.

grlee77 commented on pywt/_extensions/_pywt.pyx:
```diff
@@ -555,13 +600,15 @@ cdef public class Wavelet [type WaveletType, object WaveletObject]:
-            u"  Symmetry:       %s" % self.symmetry
+            u"  Symmetry:       %s" % self.symmetry,
+            u"  DWT:            True",
+            u"  CWT:            True"
```
typo?  CWT:  False

holgern replied:
yes typo

holgern commented:
I found an alternative data source
http://www.cpc.ncep.noaa.gov/data/indices/ersst4.nino.mth.81-10.ascii

This dataset is public domain, it can be read here: http://www.weather.gov/disclaimer under Use of NOAA/NWS Data and Products.

grlee77 commented on pywt/data/_readers.py:
```diff
@@ -141,9 +141,12 @@ def ecg():
-
+    This data contains the averaged monthly sea surface temperature in degrees 
+    Celcius of the Pacific Ocean, between 0-10 degrees South and 90-80 degrees West, from 1950 to 2016. 
+    This dataset was obtained from NOAA.
```
LGTM.  only suggestion would be to specifically mention "public domain" here
e.g.
"This dataset is in the public domain and was obtained from NOAA."

grlee77 commented:
@holgern
Great news on the data license!

merging #193 caused this to need a rebase again.  Let me know if you need any help with that.

grlee77 commented on pywt/tests/data/generate_matlab_data_cwt.py:
```diff
@@ -0,0 +1,86 @@
+size_set = 'reduced'
+
+# list of mode names in pywt and matlab
+modes = [('zero', 'zpd'),
```
I think this list of modes is not used in the CWT tests and can be removed

grlee77 commented:
Going to finally merge this one.  Thanks for sticking with it through all the revisions @holgern!

rgommers commented:
Excellent! Thanks @holgern, all.
